### PR TITLE
Enhancement - Development Environment Setup

### PR DESCRIPTION
**Pull request by a7madM**

### CODE REVIEW

File: docker-compose-dev.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+services:
+  db:
+    image: "postgres:12.2"
+    restart: always
+    env_file:
+      - src/.env
+    ports:
+      - '5432:5432'
+    volumes:
+      - /usr/local/var/postgres:/var/lib/postgresql/data
```
Comment (ahmedash95) [inline docker-compose-dev.yml:11]: would suggest to keep it local and add it to `.gitignore`

```
./postgres_data:/var/lib/postgresql/data
```

Comment (a7madM) [reply docker-compose-dev.yml:11]: why ?

**Comment by a7madM** [comment]
@emad-elsaid 
all checks have passed ya عماااااااااد

File: .github/workflows/ruby.yml [hunk 2]
```diff
@@ -29,6 +32,7 @@ jobs:
         run: |
           sudo apt-get -yqq install libpq-dev
           gem install bundler
+          cd src
```
Comment (emad-elsaid) [inline .github/workflows/ruby.yml:35]: we can use instead? also in the `yarn` step?
```
working-directory: ./src
```

File: docker-compose-dev.yml (hunk 1 above)
Comment (emad-elsaid) [reply docker-compose-dev.yml:11]: [R] lets use a named volume here, writing to a directory assumes this path isn't occupied on the machine by another setup and will just bloat the desk, using a volume will let you clean the unused volumes with `docker purge` when needed.
```suggestion
      - rubyfunctions_db:/var/lib/postgresql/data
```

File: src/.env.sample [hunk 3]
```diff
@@ -11,3 +11,8 @@ RAILS_LOG_TO_STDOUT=true
 RAILS_ENV=
 SECRET_KEY_BASE=
 SENTRY_DSN=
+
+POSTGRES_USER=postgres
+POSTGRES_PASSWORD=root
+POSTGRES_HOST=db
+POSTGRES_PORT=5432
```
Comment (emad-elsaid) [inline src/.env.sample:18]: [R] we'll need to remove all the values from the .env.sample for be consistent with the rest of the file

File: src/Gemfile [hunk 4]
```diff
@@ -24,8 +24,8 @@ group :development do
 end
 
 group :test do
-  gem 'faker'
-  gem 'shoulda-matchers'
+  gem 'faker', '~> 2.11.0'
+  gem 'shoulda-matchers', '~> 4.3'
```
Comment (emad-elsaid) [inline src/Gemfile:28]: [Q] why do you need to fix the version here? it's already locked in `gemfile.lock`?

File: src/config/database.yml [hunk 5]
```diff
@@ -1,8 +1,13 @@
 default: &default
   adapter: postgresql
+  encoding: utf8
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
+  username: <%= ENV.fetch('POSTGRES_USER', 'postgres') %>
+  password: <%= ENV.fetch('POSTGRES_PASSWORD', 'root') %>
```
Comment (emad-elsaid) [inline src/config/database.yml:6]: [Q] this isn't true for the default setup of postgres
```suggestion
  password: <%= ENV['POSTGRES_PASSWORD'] %>
```

File: src/config/database.yml [hunk 6]
```diff
@@ -1,8 +1,13 @@
...
+  host: <%= ENV.fetch('POSTGRES_HOST', 'localhost') %>
+  socket: <%= ENV.fetch('POSTGRES_PORT', 5432) %> 
+  port: <%= ENV.fetch('POSTGRES_PORT', 5432) %>
```
Comment (emad-elsaid) [inline src/config/database.yml:9]: [Q] Why do we need both socket and port keys?

File: src/lib/tasks/check_db.rake [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+  desc 'Checks to see if the database exists'
+  task exists: :environment do
+    Rake::Task['environment'].invoke
+    ActiveRecord::Base.connection
+  rescue StandardError
+    exit 1
+  else
+    exit 0
+  end
+end
```
Comment (emad-elsaid) [inline src/lib/tasks/check_db.rake:11]: why do you need this file?

File: src/scripts/entrypoint.sh [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+./scripts/wait-for-it.sh -t 0 $POSTGRES_HOST:$POSTGRES_PORT
+
+kill -9 $(cat /app/tmp/pids/server.pid)
+rm -rf /app/tmp/pids/server.pid
+yarn install --check-files
+
+bundle config set with 'development test'
+bundle install
+
+bundle exec rails db:exists || bundle exec rails db:setup
```
Comment (emad-elsaid) [inline src/scripts/entrypoint.sh:11]: You can use `rails db:prepare` this is what it does
```
rails db:prepare                         # Runs setup if database does not exist, or runs migrations if it does
```

File: src/scripts/wait-for-it.sh [hunk 9]
```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env bash
```
Comment (emad-elsaid) [inline src/scripts/wait-for-it.sh:1]: [Q] Why do you need this file? if I understand correctly and you want to wait for postgres to be online, then there is probably a cleaner solution here 
https://github.com/peter-evans/docker-compose-healthcheck#waiting-for-postgresql-to-be-healthy

**Review by emad-elsaid — CHANGES_REQUESTED** [review]
Nice one :smiley: , I have some requests to change to make it slimmer,

File: .github/workflows/ruby.yml (hunk 2 above)
Comment (a7madM) [reply .github/workflows/ruby.yml:35]: where should I write this line ?

**Comment by emad-elsaid** [comment]
Thanks for your time man, that was nice, if you like you can add another section to README.md to describe how to use this development env

**Comment by a7madM** [comment]
> Thanks for your time man, that was nice, if you like you can add another section to README.md to describe how to use this development env

I will update it.
thank you Emad

**Pull request merged**